sys/net/gnrc/transport_layer/tcp/gnrc_tcp_eventloop.c

```diff
@@ -0,0 +1,259 @@
+        if (tcp == NULL) {
+            DEBUG("gnrc_tcp_eventloop : _receive() : Header marking failed\n");
+            gnrc_pktbuf_release(pkt);
+            return -ENOMSG;
+        }
+        pkt->type = GNRC_NETTYPE_UNDEF;
+    }
+
+    hdr = (tcp_hdr_t *)tcp->data;
+
+    /* Validate Checksum */
+    if (byteorder_ntohs(hdr->checksum) != _pkt_calc_csum(tcp, ip6, pkt)) {
+        DEBUG("gnrc_tcp_eventloop : _received() : Invalid checksum\n");
+        gnrc_pktbuf_release(pkt);
+        return -EINVAL;
+    }
+
+    /* De-multiplex connection */
+    syn = ((ctl & MSK_SYN_ACK) == MSK_SYN) ? true : false;
+    src = byteorder_ntohs(hdr->src_port);
+    dst = byteorder_ntohs(hdr->dst_port);
+
+    /* Iterate over tcb list to find the connections tcb */
+    mutex_lock(&_list_conn_tcp_lock);
+    conn = _list_conn_tcp_head;
+    while (conn) {
+        /* If SYN is set, a connection is listening on that port ... */
+        if (syn && conn->local_port == dst && conn->state == LISTEN) {
+            /* ... If this Paket is for the specified address or the local addr is unspec*/
+            if(ipv6_addr_equal((ipv6_addr_t *)conn->local_addr, &((ipv6_hdr_t * ) ip6->data)->dst)
+            || ipv6_addr_is_unspecified((ipv6_addr_t *)conn->local_addr)
+            ){
+                break;
+            }
+        }
+
+        /* If portnumbers and addresses match and SYN only is not set */
+        /* Note: Make addresscheck IP agnostic if IPv6 is ever implemented !!! */
+        if (!syn && conn->local_port == dst && conn->peer_port == src
+        && ipv6_addr_equal((ipv6_addr_t *)conn->peer_addr, &((ipv6_hdr_t *) ip6->data)->src)
+        ) {
+            break;
+        }
+        conn = conn->next;
+    }
+    mutex_unlock(&_list_conn_tcp_lock);
+
+    /* Call FSM with event RCVD_PKT if a fitting connection was found */
+    if (conn != NULL){
+        if (_fsm(conn, RCVD_PKT, pkt, NULL, 0) == -ENOMEM) {
+            DEBUG("gnrc_tcp_eventloop : _receive() : FSM returned -ENOMEM\n");
+        }
+    }
+    /* No fitting connection has been found. Respond with reset */
+    else {
+        DEBUG("gnrc_tcp_eventloop : _receive() : Can't find fitting connection\n");
+        if ((ctl & MSK_RST) != MSK_RST) {
+            _pkt_build_reset_from_pkt(&rst, pkt);
+            gnrc_netapi_send(_tcp_pid, rst);
+        }
+        return -ENOTCONN;
+    }
+
+    gnrc_pktbuf_release(pkt);
+    return 0;
+}
+
+void *_event_loop(__attribute__((unused)) void *arg)
+{
+    msg_t msg;
+    msg_t reply;
+    msg_t msg_queue[GNRC_TCP_MSG_QUEUE_SIZE];
+
+    /* Store pid */
+    _tcp_pid = thread_getpid();
+
+    /* setup reply message */
+    reply.type = GNRC_NETAPI_MSG_TYPE_ACK;
+    reply.content.value = (uint32_t)-ENOTSUP;
+
+    /* Init message queue*/
+    msg_init_queue(msg_queue, GNRC_TCP_MSG_QUEUE_SIZE);
+
+    /* Register GNRC_tcp in netreg */
+    gnrc_netreg_entry_t entry;
+    entry.demux_ctx = GNRC_NETREG_DEMUX_CTX_ALL;
```

Comment:
Port to #5524. Since you want to use #5526 for `sock` (to be able to exchange the sock between different threads), I would suggest to depend this PR on #5526 anyway (which includes #5524).